samza-core/src/main/scala/org/apache/samza/job/local/ProcessJob.scala

```diff
@@ -19,113 +19,118 @@
 
 package org.apache.samza.job.local
 
-import java.io.{InputStream, OutputStream}
 import java.util.concurrent.CountDownLatch
 
-import org.apache.samza.SamzaException
 import org.apache.samza.coordinator.JobModelManager
-import org.apache.samza.job.ApplicationStatus.{New, Running, UnsuccessfulFinish}
+import org.apache.samza.job.ApplicationStatus.{New, Running, SuccessfulFinish, UnsuccessfulFinish}
 import org.apache.samza.job.{ApplicationStatus, CommandBuilder, StreamJob}
 import org.apache.samza.util.Logging
 
 import scala.collection.JavaConverters._
 
-class ProcessJob(commandBuilder: CommandBuilder, jobCoordinator: JobModelManager) extends StreamJob with Logging {
-  var jobStatus: Option[ApplicationStatus] = None
-  var process: Process = null
-
-  def submit: StreamJob = {
-    jobStatus = Some(New)
-    val waitForThreadStart = new CountDownLatch(1)
+object ProcessJob {
+  private def createProcessBuilder(commandBuilder: CommandBuilder): ProcessBuilder = {
     val processBuilder = new ProcessBuilder(commandBuilder.buildCommand.split(" ").toList.asJava)
+    processBuilder.environment.putAll(commandBuilder.buildEnvironment)
+
+    // Pipe all output to this process's streams.
+    processBuilder.redirectOutput(ProcessBuilder.Redirect.INHERIT)
+    processBuilder.redirectError(ProcessBuilder.Redirect.INHERIT)
 
     processBuilder
-      .environment
-      .putAll(commandBuilder.buildEnvironment)
+  }
+}
+
+class ProcessJob(commandBuilder: CommandBuilder, val jobModelManager: JobModelManager) extends StreamJob with Logging {
+
+  import ProcessJob._
+
+  val lock = new Object
+  val processBuilder: ProcessBuilder = createProcessBuilder(commandBuilder)
+  var jobStatus: ApplicationStatus = New
+  var processThread: Option[Thread] = None
+
 
-    // create a non-daemon thread to make job runner block until the job finishes.
-    // without this, the proc dies when job runner ends.
-    val procThread = new Thread {
+  def submit: StreamJob = {
```

Comment:
Do we need to explicitly lock submit() and kill() methods as well?

If two different threads call submit and kill at the same time, depending upon execution order and timing we could get different results(in worst case could corrupt state).

Reply:
This is an astute observation. 

My intent/understanding is that `ProcessJob` itself is not thread-safe, i.e. we do not expect callers to invoke `submit` or `kill` concurrently. Callers are expected to invoke the methods on this class serially. 

We only need to employ locking on the methods that deal with shared state between `processThread` and the thread invoking `ProcessJob` methods, i.e. `getStatus`, `setStatus`, `waitForStatus`.